autograd.h

```diff
@@ -382,24 +393,118 @@ AUTOGRAD_OPTIMIZER_CLASS(SGD) {
   AUTOGRAD_KWARG(SGD, double, weight_decay, 0, 0);
   AUTOGRAD_KWARG(SGD, bool, nesterov, false, true);
   void step() override;
+  void init_state() override;
+
+  template <class Archive>
+  void serialize(Archive & ar) { 
+    ar(CEREAL_NVP(momentum_buffers_)); 
+  }
   
+ private:
+  friend class cereal::access;
+  SGD() { }
   double lr_;
-  std::unordered_map<std::string, at::Tensor> momentum_buffers;
+  std::unordered_map<std::string, at::Tensor> momentum_buffers_;
 };
 
 AUTOGRAD_OPTIMIZER_CLASS(Adam) {
  public:
+  using OptimizerImpl::OptimizerImpl;
   Adam(Container model, double lr) : Optimizer_CRTP(model), lr_(lr) { }
   AUTOGRAD_KWARG(Adam, double, beta1, 0.9, 0.9);
   AUTOGRAD_KWARG(Adam, double, beta2, 0.999, 0.999);
   AUTOGRAD_KWARG(Adam, double, weight_decay, 0, 0);
   AUTOGRAD_KWARG(Adam, double, eps, 1e-8, 1e-8);
   void step() override;
+  void init_state() override;
+
+  template <class Archive>
+  void serialize(Archive & ar) { 
+    ar(CEREAL_NVP(step_buffer_),
+       CEREAL_NVP(exp_avg_buffer_),
+       CEREAL_NVP(exp_avg_sq_buffer_)); 
+  }
   
+ private:
+  friend class cereal::access;
+  Adam() { }
   double lr_;
-  std::unordered_map<std::string, int> step_buffer;
-  std::unordered_map<std::string, at::Tensor> exp_avg_buffer;
-  std::unordered_map<std::string, at::Tensor> exp_avg_sq_buffer;
+  std::unordered_map<std::string, int> step_buffer_;
+  std::unordered_map<std::string, at::Tensor> exp_avg_buffer_;
+  std::unordered_map<std::string, at::Tensor> exp_avg_sq_buffer_;
 };
 
 }  // namespace autograd
+
+// This is super ugly and I don't know how to simplify it
+CEREAL_REGISTER_TYPE(autograd::SGD);
+CEREAL_REGISTER_POLYMORPHIC_RELATION(autograd::OptimizerImpl, autograd::SGD);
+CEREAL_REGISTER_TYPE(autograd::Adam);
+CEREAL_REGISTER_POLYMORPHIC_RELATION(autograd::OptimizerImpl, autograd::Adam);
+
+namespace cereal {
+template<class Archive>
+void save(Archive & archive, at::Tensor const & tensor) { 
+  if (!tensor.defined()) {
+    auto type = at::ScalarType::Undefined;
+    archive(CEREAL_NVP(type));
+    return;
+  } else {
+    auto type = tensor.type().scalarType();
+    archive(CEREAL_NVP(type));
+  }
+  auto sizes = std::vector<int64_t>();
+  auto buf = std::vector<uint8_t>();
+  for (auto s : tensor.sizes()) {
+    sizes.push_back(s);
+  }
+  auto contig = tensor.toBackend(at::kCPU).contiguous();
+  auto size = tensor.storage()->size() * tensor.storage()->elementSize();
+  at::Backend backend = tensor.type().backend();
+
+  buf.resize(size);
+  memcpy(buf.data(), contig.storage()->data(), size);
+
+  archive(CEREAL_NVP(backend), CEREAL_NVP(sizes), CEREAL_NVP(buf)); 
```

Comment:
Possible optimization since you're dealing with binary archives only anyway?
```cpp
archive(CEREAL_NVP(backend), CEREAL_NVP(sizes));
archive.saveBinary(contig.storage()->data(), size); 
```

And actually, you could move the code into the cpp file if you're willing to support one or two archive types only.

Reply:
It's sometimes nice to have json to debug... but i guess support json seems a bit painful anyway.